Implement parser for request diffs

eduardoj opened the pull request:
### For reviewers
- In the review-app, go to [notifications, filtered by requests](https://obs-reviewlab.opensuse.org/eduardoj-featurediff_parser/my/notifications?type=requests),
- click on any of the request,
- click on the "Changes" tab,
- View the rendered diff.

krauselukas commented on src/api/Gemfile:
```diff
@@ -75,8 +75,6 @@ gem 'influxdb-rails'
 # Server-side datatables
 gem 'ajax-datatables-rails'
-# Add syntax highlight in ruby
-gem 'coderay'
```
@eduardoj I think we cannot remove coderay, this is still used in non beta and other places

krauselukas commented on src/api/Gemfile:
```diff
@@ -75,8 +75,6 @@ gem 'influxdb-rails'
 # Server-side datatables
 gem 'ajax-datatables-rails'
-# Add syntax highlight in ruby
-gem 'coderay'
```
@eduardoj I think we cannot remove coderay, this is still used in non beta and other places

eduardoj replied:
Ops, true. I restored the gem requirement. I will add code to keep using codery for rendering diffs for non-beta.

eduardoj replied:
Code add. Please take a look again.

dmarcoux commented on src/api/app/components/sourcediff_tab_component.rb:
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'action_sourcediff_parser'
```
Why is this `require` statement needed here, but not in the other view component above?

eduardoj replied:
I forgot to remove this `require` when I moved the code from `src/api/lib/` to `src/api/services/`. Removed. Thanks!

eduardoj commented on src/api/app/services/action_sourcediff_parser.rb:
```diff
@@ -0,0 +1,67 @@
+
+  def call
+    @action_sourcediff&.each do |sourcediff|
+      sourcediff['filenames']&.each do |filename|
```
Then we should move the logic of iterating through the hash from the backend to other place... outside of the service?

eduardoj replied:
Do you think that this is a needed refactoring? In that case, could we leave that for a following pull request?

eduardoj commented on src/api/app/services/action_sourcediff_parser.rb:
```diff
@@ -0,0 +1,67 @@
+
+  def parse_line(line)
+    case line
+    when /^\\ /
```
Works for me. I changed the login. Please take a look again.

danidoni commented on src/api/app/services/action_sourcediff_parser.rb:
```diff
@@ -0,0 +1,74 @@
+
+  private
+
+  def parse_file_diffs(content)
```
```suggestion
  def parse_file_diffs(content = [])
```
With this, you can remove the ampersand in `content&`

eduardoj replied:
Done. Thanks!

danidoni commented on src/api/app/services/action_sourcediff_parser.rb:
```diff
@@ -0,0 +1,74 @@
+    @last_original_index = 0
+    @last_changed_index = 0
+    lines = []
+    content&.each_line&.with_index do |line, line_index|
```
```suggestion
    content&.each_line.with_index do |line, line_index|
```
I believe this ampersand is not needed as applying the change above, you no longer receive `nil` in content.

eduardoj replied:
True. Done.

eduardoj commented on src/api/app/services/action_sourcediff_parser.rb:
```diff
@@ -0,0 +1,74 @@
+    end
+
+    state
+  end
```
True. Applied. Thanks!

dmarcoux commented on src/api/app/assets/stylesheets/webui/diff.scss:
```diff
@@ -9,3 +9,69 @@
+      .content {
+        margin-left: 1rem;
+        white-space: nowrap;
+        white-space: pre;
```
This rule is overwriting what the previous line does. Which rule should stay?

eduardoj replied:
Having a look at [the docs](https://developer.mozilla.org/en-US/docs/Web/CSS/white-space), I would say `pre`. Using `nowrap`, sequences of white space would be collapsed, and it would make more difficult to compare lines with changes.

I removed the `white-space: nowrap;` line. Please take a look again.

eduardoj commented:
Hey guys, I've refactored the code, and now the parsed data is accesible as an array of objects. I also added tests for the parser.

Please give it another try and leave a review.

eduardoj commented on src/api/app/views/webui/package/_revision_diff_detail.html.haml:
```diff
@@ -8,7 +8,42 @@
+                      = line.changed_index
+                  .value<
+                    %span.character><
+                      - case line.state
```
Same thought here. On it.

eduardoj replied:
Done. Please take a look again.

eduardoj commented on src/api/app/services/diff_parser.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def create_line(line, line_index)
+    starting_position = @state == 'comment' ? 0 : 1
```
Agree. On it.

eduardoj replied:
Done.

dmarcoux commented:
The second commit mentions specs, but there are no specs in this PR. Did you push them?

eduardoj commented:
> The second commit mentions specs, but there are no specs in this PR. Did you push them?

Now, yes. :sweat_smile:
